File: javascript/5_smallest_multiple.js


function prime_factorize(numb){

  var prime_map = {};
  for(var i=2; i<=numb; i++){
    if(numb % i == 0){
      if (i in prime_map){
        prime_map[i] += 1;
      }
      else{
        prime_map[i] = 1;
      }
      numb /= i;
      i--;
    }
  }
  return prime_map;
}


function update(total, primes){

  for(var prime in primes){
    if(primes.hasOwnProperty(prime)){
      if (prime in total){
        total[prime] = Math.max(total[prime], primes[prime]);
      }
      else{
        total[prime] = primes[prime];
      }
    }
  }
}

var el = {};
var total = {};
for(var i=2; i <= 20; i++){
  el = prime_factorize(i);
  update(total, el);
}


var total_count = 1;
for(var prime in total){
  if(total.hasOwnProperty(prime)){
    total_count *= Math.pow(prime, total[prime]);
  }
}
console.log(total_count);

exports.prime_factorize = prime_factorize

File: python/5_smallest_multiple.py
import math
from collections import defaultdict

def prime_factorize(numb):
  primes = defaultdict(int)
  i = 2
  while(i <= int(math.sqrt(numb)) + 1):
    if (numb % i == 0):
      primes[i] += 1
      numb /= i
      i -= 1
    i += 1
  if numb > 1:
    primes[numb] += 1
  return primes

def update(map1, map2):
  for elem in map2:
    map1[elem] = max(map1[elem], map2[elem])

total = defaultdict(int)
cur = defaultdict(int)
for i in xrange(2,20+1):
  cur = prime_factorize(i)
  update(total,cur)

total_count = 1
for elem in total:
  total_count *= pow(elem, total[elem])

print total_count